File: sdk/python/src/cyt_indexer/bm25_cohesion.py
"""BM25 lexical cohesion chunker (standalone, no pageindex)."""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any

from cyt_indexer._native import bm25_cohesion_chunk as _bm25_cohesion_chunk


@dataclass
class Bm25CohesionConfig:
    window_mode: str = "sentence"
    threshold: float = 0.8
    merge_threshold: float = 0.7
    chunk_size: int = 2048
    token_counter: str = "approximate"
    similarity_window: int = 3
    next_unit_size: int = 1
    skip_window: int = 0
    min_units_per_chunk: int = 1
    min_characters_per_sentence: int = 24
    min_characters_per_word: int = 2
    delimiters: list[str] = field(default_factory=lambda: [". ", "! ", "? ", "\n"])
    include_delim: str = "prev"
    use_stopwords: bool = True
    filter_window: int = 5
    filter_polyorder: int = 3
    filter_tolerance: float = 0.2
    stem_language: str = "english"

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def default_bm25_cohesion_config() -> Bm25CohesionConfig:
    return Bm25CohesionConfig()


def chunk(
    text: str,
    config: Bm25CohesionConfig | dict[str, Any] | None = None,
) -> list[dict[str, Any]]:
    """Split text into lexical-cohesion chunks using BM25 tokenization."""
    cfg = _cohesion_config_dict(config)
    return _bm25_cohesion_chunk(text, cfg)


def _cohesion_config_dict(
    config: Bm25CohesionConfig | dict[str, Any] | None,
) -> dict[str, Any] | None:
    if config is None:
        return None
    if isinstance(config, Bm25CohesionConfig):
        return config.to_dict()
    return config


# Public alias for app/SDK config builders.
cohesion_config_dict = _cohesion_config_dict

File: sdk/typescript/src/bm25Cohesion.ts
/** BM25 lexical cohesion chunker (standalone, no pageindex). */

import { bm25CohesionChunkNative } from "./native.js";

export interface Bm25CohesionConfig {
  windowMode?: "sentence" | "word";
  threshold?: number;
  mergeThreshold?: number;
  chunkSize?: number;
  tokenCounter?: "approximate" | "character" | "tiktoken";
  similarityWindow?: number;
  nextUnitSize?: number;
  skipWindow?: number;
  minUnitsPerChunk?: number;
  minimumWords?: number;
  minimumSentences?: number;
  minCharactersPerSentence?: number;
  minCharactersPerWord?: number;
  delimiters?: string[];
  includeDelim?: "prev" | "next";
  useStopwords?: boolean;
  filterWindow?: number;
  filterPolyorder?: number;
  filterTolerance?: number;
  stemLanguage?: string;
}

export interface CohesionChunk {
  text: string;
  start_index: number;
  end_index: number;
  token_count: number;
}

export function defaultBm25CohesionConfig(): Bm25CohesionConfig {
  return {
    windowMode: "sentence",
    threshold: 0.8,
    mergeThreshold: 0.7,
    chunkSize: 2048,
    tokenCounter: "tiktoken",
    similarityWindow: 3,
    nextUnitSize: 1,
    skipWindow: 0,
    minUnitsPerChunk: 1,
    minimumWords: 10,
    minimumSentences: 1,
    minCharactersPerSentence: 24,
    minCharactersPerWord: 2,
    delimiters: [". ", "! ", "? ", "\n"],
    includeDelim: "prev",
    useStopwords: true,
    filterWindow: 5,
    filterPolyorder: 3,
    filterTolerance: 0.2,
    stemLanguage: "english",
  };
}

/** Map camelCase SDK config to snake_case keys for the native bridge. */
export function cohesionConfigToNative(
  config?: Partial<Bm25CohesionConfig>,
): Record<string, unknown> | undefined {
  if (!config) return undefined;
  const out: Record<string, unknown> = {};
  if (config.windowMode !== undefined) out.window_mode = config.windowMode;
  if (config.threshold !== undefined) out.threshold = config.threshold;
  if (config.mergeThreshold !== undefined)
    out.merge_threshold = config.mergeThreshold;
  if (config.chunkSize !== undefined) out.chunk_size = config.chunkSize;
  if (config.tokenCounter !== undefined)
    out.token_counter = config.tokenCounter;
  if (config.similarityWindow !== undefined)
    out.similarity_window = config.similarityWindow;
  if (config.nextUnitSize !== undefined)
    out.next_unit_size = config.nextUnitSize;
  if (config.skipWindow !== undefined) out.skip_window = config.skipWindow;
  if (config.minUnitsPerChunk !== undefined)
    out.min_units_per_chunk = config.minUnitsPerChunk;
  if (config.minimumWords !== undefined)
    out.minimum_words = config.minimumWords;
  if (config.minimumSentences !== undefined)
    out.minimum_sentences = config.minimumSentences;
  if (config.minCharactersPerSentence !== undefined) {
    out.min_characters_per_sentence = config.minCharactersPerSentence;
  }
  if (config.minCharactersPerWord !== undefined) {
    out.min_characters_per_word = config.minCharactersPerWord;
  }
  if (config.delimiters !== undefined) out.delimiters = config.delimiters;
  if (config.includeDelim !== undefined)
    out.include_delim = config.includeDelim;
  if (config.useStopwords !== undefined)
    out.use_stopwords = config.useStopwords;
  if (config.filterWindow !== undefined)
    out.filter_window = config.filterWindow;
  if (config.filterPolyorder !== undefined)
    out.filter_polyorder = config.filterPolyorder;
  if (config.filterTolerance !== undefined)
    out.filter_tolerance = config.filterTolerance;
  if (config.stemLanguage !== undefined)
    out.stem_language = config.stemLanguage;
  return out;
}

export function bm25CohesionChunk(
  text: string,
  config?: Partial<Bm25CohesionConfig>,
): CohesionChunk[] {
  return bm25CohesionChunkNative(
    text,
    cohesionConfigToNative(config),
  ) as CohesionChunk[];
}
